File: infra/src/test/java/com/beat/infra/persistence/promotion/mapper/PromotionPersistenceMapperTest.java
package com.beat.infra.persistence.promotion.mapper;

import static org.junit.jupiter.api.Assertions.assertAll;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

import org.junit.jupiter.api.Test;

import com.beat.domain.promotion.domain.CarouselNumber;
import com.beat.domain.promotion.domain.Promotion;
import com.beat.infra.persistence.promotion.entity.PromotionJpaEntity;

class PromotionPersistenceMapperTest {

	private final PromotionPersistenceMapper mapper = new PromotionPersistenceMapper();

	@Test
	void toDomainPreservesJpaEntityFieldsUsedByJavaCallers() {
		PromotionJpaEntity entity = PromotionJpaEntity.rehydrate(
			11L,
			"https://example.com/promotion.png",
			22L,
			"https://example.com/performance",
			true,
			CarouselNumber.THREE
		);

		Promotion promotion = mapper.toDomain(entity);

		assertAll(
			() -> assertEquals(11L, promotion.getId()),
			() -> assertEquals("https://example.com/promotion.png", promotion.getPromotionPhoto()),
			() -> assertEquals(22L, promotion.getPerformanceId()),
			() -> assertEquals("https://example.com/performance", promotion.getRedirectUrl()),
			() -> assertTrue(promotion.isExternal()),
			() -> assertEquals(CarouselNumber.THREE, promotion.getCarouselNumber())
		);
	}

	@Test
	void toEntityPreservesDomainFieldsAndJavaVisiblePromotionJpaEntityContract() {
		Promotion promotion = Promotion.rehydrate(
			31L,
			"https://example.com/internal.png",
			null,
			"/notices/31",
			false,
			CarouselNumber.ONE
		);

		PromotionJpaEntity entity = mapper.toEntity(promotion);

		assertAll(
			() -> assertEquals(31L, entity.getId()),
			() -> assertEquals("https://example.com/internal.png", entity.getPromotionPhoto()),
			() -> assertNull(entity.getPerformanceId()),
			() -> assertEquals("/notices/31", entity.getRedirectUrl()),
			() -> assertFalse(entity.isExternal()),
			() -> assertEquals(CarouselNumber.ONE, entity.getCarouselNumber())
		);
	}
}

File: infrastructure/src/test/kotlin/com/beat/infrastructure/persistence/promotion/mapper/PromotionPersistenceMapperTest.kt
package com.beat.infrastructure.persistence.promotion.mapper

import com.beat.domain.promotion.model.CarouselNumber
import com.beat.domain.promotion.model.Promotion
import com.beat.infrastructure.persistence.promotion.entity.PromotionJpaEntity
import io.kotest.core.spec.style.FunSpec
import io.kotest.matchers.shouldBe

class PromotionPersistenceMapperTest : FunSpec({
    val mapper = PromotionPersistenceMapper()

    test("toDomain은 JPA entity 필드를 보존한다") {
        val entity = PromotionJpaEntity.rehydrate(
            11L,
            "https://example.com/promotion.png",
            22L,
            "https://example.com/performance",
            true,
            CarouselNumber.THREE,
        )

        val promotion = mapper.toDomain(entity)

        promotion.id shouldBe 11L
        promotion.promotionPhoto shouldBe "https://example.com/promotion.png"
        promotion.performanceId shouldBe 22L
        promotion.redirectUrl shouldBe "https://example.com/performance"
        promotion.isExternal shouldBe true
        promotion.carouselNumber shouldBe CarouselNumber.THREE
    }

    test("toEntity는 신규 Promotion의 생성 id를 null로 유지한다") {
        val promotion = Promotion.create(
            "https://example.com/new.png",
            44L,
            "https://example.com/new-performance",
            true,
            CarouselNumber.TWO,
        )

        val entity = mapper.toEntity(promotion)

        promotion.id shouldBe null
        entity.id shouldBe null
        entity.performanceId shouldBe 44L
        entity.carouselNumber shouldBe CarouselNumber.TWO
    }

    test("toEntity는 domain 필드를 보존한다") {
        val promotion = Promotion.rehydrate(
            31L,
            "https://example.com/internal.png",
            null,
            "/notices/31",
            false,
            CarouselNumber.ONE,
        )

        val entity = mapper.toEntity(promotion)

        entity.id shouldBe 31L
        entity.promotionPhoto shouldBe "https://example.com/internal.png"
        entity.performanceId shouldBe null
        entity.redirectUrl shouldBe "/notices/31"
        entity.isExternal shouldBe false
        entity.carouselNumber shouldBe CarouselNumber.ONE
    }
})
